lib/improver/utilities/cube_extraction.py

```diff
@@ -0,0 +1,119 @@
+
+from ast import literal_eval
+import iris
+
+
+def parse_constraint_list(constraints, units):
+    """
+    Takes a list of string constraints and converts to key-value pairs
+
+    Args:
+        constraints (list):
+            Space separated list of constraints with no space between key and
+            value in each pair: e.g: kw1=val1 kw2=val2 kw3=val3.  Values must
+            be of scalar types interpretable by ast.literal_eval: strings,
+            numbers, booleans, or "None".
+        units (list of strings or None):
+            Space separated list of units for each coordinate in the list of
+            constraints.  One or more "units" may be None, and units can only
+            be associated with coordinate constraints.
+
+    Returns:
+        constraints_dict (dictionary):
+            A dictionary of constraint keys and interpreted values
+        units_dict (dictionary or None):
+            A dictionary of unit keys and values
+    """
+
+    if units is None:
+        list_units = len(constraints)*[None]
+        units_dict = None
+    else:
+        if len(units) != len(constraints):
+            msg = 'units list must match constraints'
+            raise ValueError(msg)
+        list_units = units
+        units_dict = {}
+
+    constraints_dict = {}
+    for constraint_pair, units in zip(constraints, list_units):
+        [key, value] = constraint_pair.split('=')
+        try:
+            constraints_dict[key] = literal_eval(value)
+        except ValueError:
+            constraints_dict[key] = value
+
+        if units is not None and units != 'None':
+            units_dict[key] = units
+
+    return constraints_dict, units_dict
+
+
+def extract_subcube(input_filename, constraints, units):
+    """
+    Using a set of constraints, extract a subcube from the provided cube or
+    cubelist if it is available.  Raises ValueError on merge if no subcube
+    matched the constraints provided.
+
+    Args:
+        cube (iris.cube.Cube or iris.cube.CubeList):
+            The cube or cubelist from which a subcube is to be extracted.
+        constraints (dictionary):
+            A dictionary of constraints that define the subcube to be
+            extracted.
+
+    Kwargs:
+        units (dictionary):
+            A dictionary of units for the constraints.  Supplied if any
+            coordinate constraints are provided in different units from those
+            of the input cube (eg precip in mm/h for cube threshold in m/s).
+
+    Returns:
+        cube (iris.cube.Cube):
+            A single cube matching the input constraints.
+
+    """
+    constraint = iris.Constraint(**constraints)
+
+    if units is not None:
+        cubes = iris.load(input_filename)
+        for cube in cubes:
+            for coord in units.keys():
+                cube.coord(coord).convert_units(units[coord])
+        cubes = cubes.extract(constraint)
+    else:
+        cubes = iris.load(input_filename, constraint)
```

Comment:
We should (if possible) use the centralised improver `load_cube` here (and above).
Since we do `iris.load` unconditionally, we should factor this out and move it above, too.

Reply:
Yes, we should be using the centralised load_cube wrapper here.